tools/src/bin/fap-lld/fastrel.rs

```diff
@@ -0,0 +1,156 @@
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Convert the relocations into `.fast.rel` sections.
+    let (symtab, strtab) = fap.symbol_table()?.ok_or(Error::NoSymbolTable)?;
+    let fastrel_sections = rel_sections
+        .into_iter()
+        .map(|(section_name, section)| FastRelSection::new(section_name, section, &symtab, &strtab))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Write the `.fast.rel` sections into the binary.
+    let temp_dir = tempdir()?;
+    for section in fastrel_sections {
+        let fastrel_section_name = hex::encode(Md5::digest(&section.name));
+        let data_path = temp_dir
+            .path()
+            .join(fastrel_section_name)
+            .with_extension("bin");
+        let mut data = fs::OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(&data_path)?;
+        section.write(&mut data)?;
+        data.flush()?;
+
+        let res = Command::new(objcopy)
+            .arg("--add-section")
+            .arg(format!("{}={}", section.name, data_path.display()))
+            .arg(output_fap)
+            .status()?;
+        if !res.success() {
+            return Err(Error::ObjcopyFailed);
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct FastRel<'data> {
+    section_index: u16,
+    section_value: u64,
+    r_type: u32,
+    name: &'data str,
+}
+
+impl<'data> FastRel<'data> {
+    fn gnu_sym_hash(&self) -> u32 {
+        let mut h = 0x1505;
+        for c in self.name.as_bytes() {
+            h = (h << 5) + h + u32::from(*c);
+        }
+        h
+    }
+}
+
+/// A `.fast.rel` section.
+#[derive(Debug)]
+struct FastRelSection<'data> {
+    name: String,
+    fastrel_offsets: HashMap<FastRel<'data>, Vec<u64>>,
+}
+
+impl<'data> FastRelSection<'data> {
+    fn new(
+        section_name: &str,
+        section: RelIterator<'_, LittleEndian>,
+        symtab: &SymbolTable<'_, LittleEndian>,
+        strtab: &StringTable<'data>,
+    ) -> Result<Self, Error> {
+        assert!(section_name.starts_with(".rel"));
+
+        let mut fastrel_offsets = HashMap::<_, Vec<u64>>::new();
+        for rel in section {
+            let symbol = symtab.get(rel.r_sym as usize)?;
+            let name = if symbol.st_name == 0 {
+                ""
+            } else {
+                strtab.get(symbol.st_name as usize)?
+            };
+
+            fastrel_offsets
+                .entry(FastRel {
+                    section_index: symbol.st_shndx,
+                    section_value: symbol.st_value,
+                    r_type: rel.r_type,
+                    name,
+                })
+                .or_default()
+                .push(rel.r_offset);
+        }
+
+        Ok(FastRelSection {
+            name: format!(".fast{}", section_name),
+            fastrel_offsets,
+        })
+    }
+
+    fn write(&self, mut w: impl Write) -> io::Result<()> {
+        const VERSION: u8 = 1;
```

Comment:
It might be useful to pull this out to the module level, just so it's very clear what version this module supports.